Updated code_utils.py & local_commandline_code_executor.py (powershell to pwsh)

abhaymathur21 opened the pull request:
Updated the 'powershell' command to 'pwsh' for a ps1 executable file 

<!-- Thank you for your contribution! Please review https://microsoft.github.io/autogen/docs/Contribute before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

I was trying to use autogen to execute terminal commands (eg: git status, ls) but it was throwing an error. So these changes are required if terminal commands are to be executed in the new 7.x version of PowerShell. To execute terminal commands in powershell, autogen first makes a ps1 shell script with those commands written in it and then executes that ps1 shell script. In order for that shell script execution to work, these changes are required.

<!-- Please give a short summary of the change and the problem this solves. -->

In March 2020, microsoft released a new version of powershell (7.0) in which the powershell.exe file was renamed to pwsh.exe 
<img width="451" alt="image" src="https://github.com/microsoft/autogen/assets/102252920/cb16fe8c-88cc-4e13-b3ab-e445d2b2208f">
https://learn.microsoft.com/en-us/powershell/scripting/whats-new/differences-from-windows-powershell?view=powershell-7.4

In your code, if the user is working in a powershell window and a '.ps1' script is to be executed, your code uses the 'powershell' command to execute it which has now been replaced with 'pwsh' in the new Microsoft PowerShell version 7.0
<img width="426" alt="image" src="https://github.com/microsoft/autogen/assets/102252920/6162ecf8-b7fc-45d9-b7da-1192e6b20bbc">
<img width="408" alt="image" src="https://github.com/microsoft/autogen/assets/102252920/0c15298d-f360-4fb8-b821-5695f8bb2b32">


So in your code I replaced the command 'powershell' with 'pwsh' when a ps1 script has to be executed

I also added the 'pwsh' command to the list of commands in the local_commandline_code_executor.py file 
<img width="280" alt="image" src="https://github.com/microsoft/autogen/assets/102252920/8557fb06-c2ce-45b9-b1de-e74d74848d3f">

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [x] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [x] I've made sure all auto checks have passed.

ekzhu commented on autogen/code_utils.py:
```diff
@@ -219,7 +219,7 @@ def _cmd(lang):
         return "sh"
     if lang in ["ps1"]:
-        return "powershell"
+        return "pwsh"
```
Shouldn’t the first condition need to be split to also handle “powershell”?

abhaymathur21 replied:
yes youre right, thanks for pointing that out, i have added a split condition to deal with powershell as well could you please review that once as well?

abhaymathur21 replied:
I added the pwsh command to the local commandline executor file list of commands as well, please review that once too

ekzhu reviewed (changes requested):
Thanks for the PR! Have you also checked out the new code executor? It’s here https://github.com/microsoft/autogen/blob/main/autogen/coding/local_commandline_code_executor.py 

You can use the new code executor like

```python
proxy = CoversableAgent(“proxy”, code_execution_config={“executor”: “commadline-local”})
```
Perhaps you want to add the handling for powershell there as well? It’s okay to do it in a separate PR.

abhaymathur21 commented:
@microsoft-github-policy-service agree

ekzhu commented on autogen/coding/local_commandline_code_executor.py:
```diff
@@ -128,7 +128,7 @@ def execute_code_blocks(self, code_blocks: List[CodeBlock]) -> CommandlineCodeRe
             filename_uuid = uuid.uuid4().hex
             filename = None
-            if lang in ["bash", "shell", "sh"]:
+            if lang in ["bash", "shell", "sh", "pwsh"]:
```
```suggestion
            if lang in ["bash", "shell", "sh", "pwsh", "powershell", "ps1"]:
```

ekzhu replied:
The `lang` here is only used for filename, which doesn't affect how it is executed. It uses `code_utils` later so the execution part is already handled.

abhaymathur21 replied:
oh right okay

sonichi commented on autogen/code_utils.py:
```diff
@@ -214,12 +214,14 @@ def timeout_handler(signum, frame):
     if lang in ["shell"]:
         return "sh"
     if lang in ["ps1"]:
-        return "powershell"
+        return "pwsh"
+    if lang in ["powershell"]:
```
why not merge the conditions in line 221 and 223?
Can we add a simple test to cover the if branch?

abhaymathur21 replied:
But then if lang = powershell we wont have any case to deal with that right? For the users that still use old powershell versions. Is there any other way we can cover that if branch with a test that I am unaware of at the moment?

sonichi replied:
I mean changing line 221 into: `if lang in ["ps1", "powershell"]:`

abhaymathur21 replied:
Ohh okay got it yes you're right that would be more efficient, sorry I'll make the changes right away

sonichi commented:
@ekzhu @abhaymathur21 have you tested it in windows? When I run pwsh in windows, it can't find the command. powershell is available. I think this change will break for existing users. We'd better check whether pwsh is available in the current system environment.

abhaymathur21 commented:
> @ekzhu @abhaymathur21 have you tested it in windows? When I run pwsh in windows, it can't find the command. powershell is available. I think this change will break for existing users. We'd better check whether pwsh is available in the current system environment.


<img width="852" alt="image" src="https://github.com/microsoft/autogen/assets/102252920/5edc4a61-3b97-4f7a-863a-be234326a64c">

I have tested it in windows yes. I was getting an error that's why I did some research and found out that since mine is a relatively new laptop, it came with the 7.0 PowerShell installed so the command that works for me is 'pwsh'. But for users that still operate 5.0 PowerShell the command is 'powershell'. So is there any way we can customize this command for the user in our code based on the version of powershell they are using maybe?

sonichi commented:
I asked on Discord. I prefer detecting the availability or making it configurable.

abhaymathur21 commented:
> I asked on Discord. I prefer detecting the availability or making it configurable.

I added a function to detect the command (powershell or pwsh) which works in the user's system and then use that command only
<img width="618" alt="image" src="https://github.com/microsoft/autogen/assets/102252920/3b5154ef-cb90-4e34-b811-96b1336ad60d">
<img width="427" alt="image" src="https://github.com/microsoft/autogen/assets/102252920/6337c186-5bc2-4377-b9a2-216528a718b2">

Is this okay? I tested it, it works with autogen on my PC

sonichi commented:
That works. Could you add a unit test to cover the new function?

abhaymathur21 commented:
> That works. Could you add a unit test to cover the new function?

Sure, I wrote this unit test and added it to your test_code_utils.py (which was incorrectly named as test_code.py initially but I corrected it to test_code_utils.py using my own discretion):

class TestGetPowerShellCommand(unittest.TestCase):
    
    @patch('subprocess.run')
    def test_get_powershell_command_powershell(self, mock_subprocess_run):
        # Set up the mock to return a successful result for 'powershell'
        mock_subprocess_run.return_value.returncode = 0
        mock_subprocess_run.return_value.stdout = StringIO("5")
        
        self.assertEqual(get_powershell_command(), "powershell")
    
    @patch('subprocess.run')
    def test_get_powershell_command_pwsh(self, mock_subprocess_run):
        # Set up the mock to return a successful result for 'pwsh'
        mock_subprocess_run.side_effect = [FileNotFoundError, mock_subprocess_run.return_value]
        mock_subprocess_run.return_value.returncode = 0
        mock_subprocess_run.return_value.stdout = StringIO("7")
        
        self.assertEqual(get_powershell_command(), "pwsh")
    
    @patch('subprocess.run')
    def test_get_powershell_command_no_shell(self, mock_subprocess_run):
        # Set up the mock to simulate 'powershell' and 'pwsh' not found
        mock_subprocess_run.side_effect = [FileNotFoundError, FileNotFoundError]
        
        with patch('sys.stdout', new=StringIO()) as fake_out:
            get_powershell_command()
            self.assertEqual(fake_out.getvalue().strip(), "Neither powershell nor pwsh is installed.")
    
    @patch('subprocess.run')
    def test_get_powershell_command_no_shell_no_output(self, mock_subprocess_run):
        # Set up the mock to simulate 'powershell' and 'pwsh' not found without printing error message
        mock_subprocess_run.side_effect = [FileNotFoundError, FileNotFoundError]
        
        self.assertIsNone(get_powershell_command())
        
if __name__ == '__main__':
    unittest.main()

sonichi commented:
Thanks. That works. The conflict needs to be resolved though before merging.

abhaymathur21 commented:
> Thanks. That works. The conflict needs to be resolved though before merging.

Done!

sonichi commented on autogen/code_utils.py:
```diff
@@ -213,13 +213,37 @@ def timeout_handler(signum, frame):
         return lang
     if lang in ["shell"]:
         return "sh"
     if lang in ["ps1"]:
```
```suggestion
    if lang in ["ps1", "pwsh", "powershell"]:
```
Let's add alternative strs here because the language model may still suggest them.

abhaymathur21 replied:
oh yes thats a good idea
